File: .claude/hooks/file-to-markdown.mjs
#!/usr/bin/env node
// @hookstack pre-read-file-to-markdown
// Convertit PDF/DOCX/PPTX et autres fichiers binaires en Markdown avant lecture (PreToolUse Read)
import { readFileSync, existsSync } from 'fs';
import { execSync } from 'child_process';
import { fileURLToPath } from 'url';
import { extname, basename } from 'path';

const MAX_CHARS = 50_000;

const SUPPORTED = new Set(['pdf', 'docx', 'pptx', 'odt', 'rtf', 'doc', 'ppt', 'xlsx', 'epub', 'html', 'htm']);

function defaultExec(cmd) {
  return execSync(cmd, { encoding: 'utf8', timeout: 30_000 }).trim();
}

function hasBinary(name, exec) {
  try { exec(`which ${name}`); return true; } catch { return false; }
}

export function run(input, { exec = defaultExec, exists = existsSync } = {}) {
  if (input.tool_name !== 'Read') return null;

  const filePath = input.tool_input?.file_path ?? '';
  if (!filePath) return null;

  const ext = extname(filePath).toLowerCase().replace('.', '');
  if (!SUPPORTED.has(ext)) return null;

  if (!exists(filePath)) return null;

  const hasPdftotext = ext === 'pdf' && hasBinary('pdftotext', exec);
  const hasPandoc = hasBinary('pandoc', exec);

  if (!hasPdftotext && !hasPandoc) return null;

  let markdown;
  try {
    if (ext === 'pdf' && hasPdftotext) {
      markdown = exec(`pdftotext "${filePath}" -`);
    } else if (hasPandoc) {
      markdown = exec(`pandoc --to markdown --wrap=none "${filePath}"`);
    } else {
      return null;
    }
  } catch {
    return null;
  }

  if (!markdown?.trim()) return null;

  let content = markdown.trim();
  let truncated = false;
  if (content.length > MAX_CHARS) {
    content = content.slice(0, MAX_CHARS);
    truncated = true;
  }

  const name = basename(filePath);
  const suffix = truncated ? ` (truncated to ${MAX_CHARS} chars)` : '';
  const header = `[file-to-markdown] \`${name}\` converted to Markdown${suffix}:\n\n`;

  return { decision: 'block', reason: header + content };
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/file-to-markdown.py
#!/usr/bin/env python3
# @hookstack pre-read-file-to-markdown
"""Converts PDF/DOCX/PPTX and other binaries to Markdown before reading (PreToolUse Read)."""
import json
import os
import subprocess
import sys

MAX_CHARS = 50_000

SUPPORTED = {
    "pdf",
    "docx",
    "pptx",
    "odt",
    "rtf",
    "doc",
    "ppt",
    "xlsx",
    "epub",
    "html",
    "htm",
}


def _exec(cmd):
    return subprocess.run(
        cmd, shell=True, capture_output=True, text=True, timeout=30, check=True
    ).stdout.strip()


def _exists(path):
    return os.path.exists(path)


def _has_binary(name, exec_cmd):
    try:
        exec_cmd(f"which {name}")
        return True
    except Exception:
        return False


def run(input_data, *, exec_cmd=None, exists=None):
    if exec_cmd is None:
        exec_cmd = _exec
    if exists is None:
        exists = _exists

    if input_data.get("tool_name") != "Read":
        return None

    file_path = (input_data.get("tool_input") or {}).get("file_path") or ""
    if not file_path:
        return None

    ext = os.path.splitext(file_path)[1].lower().lstrip(".")
    if ext not in SUPPORTED:
        return None

    if not exists(file_path):
        return None

    has_pdftotext = ext == "pdf" and _has_binary("pdftotext", exec_cmd)
    has_pandoc = _has_binary("pandoc", exec_cmd)

    if not has_pdftotext and not has_pandoc:
        return None

    try:
        if ext == "pdf" and has_pdftotext:
            markdown = exec_cmd(f'pdftotext "{file_path}" -')
        elif has_pandoc:
            markdown = exec_cmd(f'pandoc --to markdown --wrap=none "{file_path}"')
        else:
            return None
    except Exception:
        return None

    if not (markdown or "").strip():
        return None

    content = markdown.strip()
    truncated = False
    if len(content) > MAX_CHARS:
        content = content[:MAX_CHARS]
        truncated = True

    name = os.path.basename(file_path)
    suffix = f" (truncated to {MAX_CHARS} chars)" if truncated else ""
    header = f"[file-to-markdown] `{name}` converted to Markdown{suffix}:\n\n"

    return {"decision": "block", "reason": header + content}


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
